airbyte-integrations/connectors/source-snapchat-marketing/source_snapchat_marketing/components.py

```diff
@@ -0,0 +1,103 @@
+#
+# Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+#
+
+import datetime
+from dataclasses import dataclass
+from typing import List, Optional, Union
+
+import dpath
+from airbyte_cdk.sources.declarative.incremental import DatetimeBasedCursor
+from airbyte_cdk.sources.declarative.transformations import AddFields
+from airbyte_cdk.sources.declarative.types import StreamSlice, StreamState
+from airbyte_cdk.sources.types import Config, Record, StreamSlice, StreamState
+from isodate import Duration, parse_duration
+
+
+@dataclass
+class SnapchatMarketingDatetimeBasedCursor(DatetimeBasedCursor):
+    """
+    SnapchatMarketingDatetimeBasedCursor is a variant of DatetimeBasedCursor that strictly compares start and end dates.
+    It ensures that the start date is always less than the end date to prevent errors.
```

Comment:
is this really the only difference?

can you clarify what errors are prevented why we're not adding this check in the CDK instead of writing a custom component?

Reply:
Yes, this is the only difference. This change is necessary to ensure that we don't request data with the same start_time and end_time, as the API returns an error in such cases.

For example, if we set start_date to 2024-05-17 and end_date to 2024-06-17 with the current DatetimeBasedCursor implementation, we will sync with two slices: `{'start_time': '2024-05-17T00:00:00', 'end_time': '2024-06-17T00:00:00'}` and `{'start_time': '2024-06-17T00:00:00', 'end_time': '2024-06-17T00:00:00'}` for `*_daily_analytics streams`, as this stream has a granularity of one month. We will encounter an error when trying to request data with the second slice.


<img width="753" alt="image" src="https://github.com/airbytehq/airbyte/assets/53845333/2088efba-359e-46c0-84ae-4264cf373662">

Reply:
To speed up the migration process, I have decided to leave it as a custom component. If necessary, we can add it to the CDK in a follow-up change.

Reply:
The `_partition_daterange` method is private so I would definitely prefer we update the CDK to avoid surprises.

It should only be a matter of adding a 
`inclusive_end_datetime` field to the `DatetimeBasedCursor`, which would default to true

Reply:
Updated to last low code CDK update - https://github.com/airbytehq/airbyte/pull/39603